### PR TITLE
@mczernek/app loader migration

### PR DESCRIPTION
**Pull request by mczernek**
# Why

If an application is being ejected, and some user updates its previously standalone (managed) application to ejected one, crash is observed when attempting to use AppLoader. Additionally, simple fix to prevent the crash causes helps, but tasks are not being transferred between versions correctly.

# How

Root cause of the problem is the fact, that name of explicit AppLoader class implementation is persisted in SharedPreferences. At the same time, it differs between managed and bare workflow.
Current implementation defines names of loader in meta-data. This should cause it to be changed accordingly when migrating from managed to bare workflow.

### CODE REVIEW

File: android/expoview/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -106,6 +106,11 @@
                 android:name="android.support.FILE_PROVIDER_PATHS"
                 android:resource="@xml/provider_paths" />
         </provider>
+
+      <meta-data
+        tools:replace="android:value"
```
Comment (sjchmiela) [inline android/expoview/src/main/AndroidManifest.xml:111]: What would happen without this `tools:replace` property?

File: packages/@unimodules/react-native-adapter/android/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -1,5 +1,11 @@
+<manifest package="org.unimodules.adapters.react"
+  xmlns:android="http://schemas.android.com/apk/res/android">
 
-<manifest package="org.unimodules.adapters.react">
+  <application>
+    <meta-data
+      android:name="org.unimodules.core.AppLoader#react-native-headless"
```
Comment (sjchmiela) [inline packages/@unimodules/react-native-adapter/android/src/main/AndroidManifest.xml:6]: Is there some convention around whether to use camelcased or snakecased property names?

**Review by sjchmiela — APPROVED** [review]
This is a great change! 👍

**Review by tsapeta — APPROVED** [review]

File: packages/@unimodules/react-native-adapter/android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (mczernek) [reply packages/@unimodules/react-native-adapter/android/src/main/AndroidManifest.xml:6]: I do not know of any.

File: android/expoview/src/main/AndroidManifest.xml (hunk 1 above)
Comment (mczernek) [reply android/expoview/src/main/AndroidManifest.xml:111]: It won't compile. We need to inform compile that we intentionally provide same meta-data twice.

File: packages/expo-task-manager/CHANGELOG.md [hunk 3]
```diff
@@ -6,4 +6,6 @@
 
 ### 🎉 New features
 
+- Upgraging and application does not cause BackgroundFetch tasks to unregister. ([#8348](https://github.com/expo/expo/pull/8438) by [@mczernek](https://github.com/mczernek))
```
Comment (tsapeta) [inline packages/expo-task-manager/CHANGELOG.md:9]:
```suggestion
- Upgrading an application does not cause `BackgroundFetch` tasks to unregister. ([#8348](https://github.com/expo/expo/pull/8438) by [@mczernek](https://github.com/mczernek))
```

Also, I think it's more a fix than a new feature 😄 
Could you check whether there are any reported issues related to background fetch that this PR may solve and link them in this PR to close them automatically?

File: packages/expo-task-manager/android/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -8,12 +8,15 @@
             <intent-filter>
                 <action android:name="expo.modules.taskManager.TaskBroadcastReceiver.INTENT_ACTION" />
                 <action android:name="android.intent.action.BOOT_COMPLETED" />
+                <action android:name="android.intent.action.MY_PACKAGE_REPLACED" />
```
Comment (tsapeta) [inline packages/expo-task-manager/android/src/main/AndroidManifest.xml:11]: ❓

File: packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskService.java [hunk 5]
```diff
@@ -64,18 +62,18 @@
...
   private static final Map<String, TaskExecutionCallback> sTaskCallbacks = new HashMap<>();
 
   public TaskService(Context context) {
     super();
     mContextRef = new WeakReference<>(context);
+    mTasksAndEventsRepository = TasksAndEventsRepository.create(context);
 
-    if (sTasksTable == null) {
-      sTasksTable = new HashMap<>();
+    if (!mTasksAndEventsRepository.tasksExists()) {
```
Comment (tsapeta) [inline packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskService.java:75]:
```suggestion
    if (!mTasksAndEventsRepository.tasksExist()) {
```

File: packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/BareTasksAndEventsRepository.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  public List<Bundle> getEvents(String appId) {
+    List<Bundle> allEvents = new LinkedList<>();
+    for(List<Bundle> value: sEvents.values()) {
+      allEvents.addAll(value);
+    }
+    return allEvents;
+  }
+
+  @Override
+  public boolean tasksExists() {
```
Comment (tsapeta) [inline packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/BareTasksAndEventsRepository.java:60]:
```suggestion
  public boolean tasksExist() {
```

File: packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/BareTasksAndEventsRepository.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+  @Override
+  public void removeEvents(String appId) {
+    sEvents.remove(appId);
+  }
+
+  @Override
+  public List<Bundle> getEvents(String appId) {
+    List<Bundle> allEvents = new LinkedList<>();
+    for(List<Bundle> value: sEvents.values()) {
```
Comment (tsapeta) [inline packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/BareTasksAndEventsRepository.java:53]:
```suggestion
    for (List<Bundle> value: sEvents.values()) {
```

File: packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/ManagedTasksAndEventsRepository.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    sEvents.remove(appId);
+  }
+
+  @Override
+  public List<Bundle> getEvents(String appId) {
+    return sEvents.get(appId);
+  }
+
+  @Override
+  public boolean tasksExists() {
```
Comment (tsapeta) [inline packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/ManagedTasksAndEventsRepository.java:52]:
```suggestion
  public boolean tasksExist() {
```

File: packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/TasksAndEventsRepository.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    public Map<String, Object> tasks;
+  }
+
+  void putEvents(String appId, List<Bundle> events);
+  void putEventForAppId(String appId, Bundle body);
+  boolean hasEvents(String appId);
+  void removeEvents(String appId);
+  List<Bundle> getEvents(String appId);
+
+  boolean tasksExists();
```
Comment (tsapeta) [inline packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/repository/TasksAndEventsRepository.java:43]:
```suggestion
  boolean tasksExist();
```

**Review by tsapeta — CHANGES_REQUESTED** [review]
Sooo many hacks there 😞 But looks like there is no better way right now 😞 

Left a few comments, but besides of them
- All those new files could use Kotlin 😄 
- Would be great if you could add some comments around those repository classes & methods and especially in places where we do these hacks

File: packages/expo-task-manager/android/src/main/AndroidManifest.xml (hunk 4 above)
Comment (mczernek) [reply packages/expo-task-manager/android/src/main/AndroidManifest.xml:11]: Currently, while updating native app all alarms gets cancelled. If we want them to survive, we need to reschedule them after reinstall. This system broadcasts informs newly installed app of the fact it is reinstalled. Effectively, we have no Alarms, just like after reboot, so we need to reschedule them in exactly the same way as after reboot.

**Review by tsapeta — APPROVED** [review]
Still one comment, but generally looks good now 👍 
Could you also update PR's title?

File: packages/expo-task-manager/CHANGELOG.md [hunk 10]
```diff
@@ -7,3 +7,5 @@
 ### 🎉 New features
 
 ### 🐛 Bug fixes
+
+- Upgraging and application does not cause BackgroundFetch tasks to unregister. ([#8348](https://github.com/expo/expo/pull/8438) by [@mczernek](https://github.com/mczernek))
```
[the remark this answers is not in the document]
Comment (tsapeta) [reply packages/expo-task-manager/CHANGELOG.md:11]: Actually we could put this changelog entry to `expo-background-fetch` as well 🤔

File: packages/expo-background-fetch/CHANGELOG.md [hunk 11]
```diff
@@ -7,3 +7,5 @@
 ### 🎉 New features
 
 ### 🐛 Bug fixes
+
+- Upgraging an application does not cause BackgroundFetch tasks to unregister. ([#8348](https://github.com/expo/expo/pull/8438) by [@mczernek](https://github.com/mczernek))
```
Comment (tsapeta) [inline packages/expo-background-fetch/CHANGELOG.md:11]:
```suggestion
- Upgrading an application does not cause `BackgroundFetch` tasks to unregister. ([#8348](https://github.com/expo/expo/pull/8438) by [@mczernek](https://github.com/mczernek))
```

File: packages/expo-task-manager/CHANGELOG.md [hunk 12]
```diff
@@ -7,3 +7,5 @@
 ### 🎉 New features
 
 ### 🐛 Bug fixes
+
+- Upgraging an application does not cause BackgroundFetch tasks to unregister. ([#8348](https://github.com/expo/expo/pull/8438) by [@mczernek](https://github.com/mczernek))
```
Comment (tsapeta) [inline packages/expo-task-manager/CHANGELOG.md:11]:
```suggestion
- Upgrading an application does not cause `BackgroundFetch` tasks to unregister. ([#8348](https://github.com/expo/expo/pull/8438) by [@mczernek](https://github.com/mczernek))
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**